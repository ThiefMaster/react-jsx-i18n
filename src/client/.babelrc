{
  "plugins": [
    "@babel/plugin-proposal-object-rest-spread",
    "@babel/plugin-proposal-class-properties"
  ],
  "presets": [
    "@babel/react",
    [
      "@babel/env",
      {
        "targets": {
          "browsers": ["last 2 versions", "firefox esr", "not ie<11", "not ie_mob>0"]
        }
      }
    ]
  ]
}
