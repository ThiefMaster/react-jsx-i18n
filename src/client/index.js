import React from 'react';
import PropTypes from 'prop-types';


export class Param extends React.Component {
    static propTypes = {
        // eslint-disable-next-line react/no-unused-prop-types
        name: PropTypes.string.isRequired,
        value: PropTypes.any.isRequired,
        children: PropTypes.string,
    };

    static defaultProps = {
        children: undefined,
    };

    render() {
        const {value, children} = this.props;
        return React.isValidElement(value) ? React.cloneElement(value, {}, children) : value;
    }
}

const getTranslatableString = (children) => {
    const usedParams = new Set();
    const items = React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
            // plain value, nothing to change here
            if (typeof child !== 'string') {
                throw new Error(`Unexpected Translate child type: ${typeof child}`);
            }
            return child;
        } else if (child.type !== Param) {
            throw new Error(`Translate child components must be of type Param; got ${child.type.name}`);
        } else {
            const {name: paramName, children: paramContent} = child.props;
            if (usedParams.has(paramName)) {
                throw new Error(`Translate params must be unique; found ${paramName} more than once`);
            } else if (
                paramContent !== undefined &&
                typeof paramContent !== 'string'
            ) {
                throw new Error(`Unexpected Param child type: ${typeof paramContent}`);
            }
            usedParams.add(paramName);
            return paramContent === undefined ? `{${paramName}}` : `{${paramName}}${paramContent}{/${paramName}}`;
        }
    });
    return items.join('').replace(/\s+/g, ' ').trim();
};


const jsonToReact = (values, component, props, ...children) => {
    /* eslint-disable react/prop-types */
    component = {Fragment: React.Fragment, Param}[component];
    if (component === Param) {
        // inject the value from the original Param in the translated string's Param
        props = {...props, value: values[props.name]};
    }
    return React.createElement(component, props, ...children.map((child) => {
        return typeof child === 'string' ? child : jsonToReact(values, ...child);
    }));
};


const renderTranslation = (translation, values) => {
    if (typeof translation === 'string') {
        return translation;
    } else {
        return jsonToReact(values, ...translation);
    }
};


const jsonToText = (component, props, ...children) => {
    if (component === 'Param') {
        if (children.length) {
            return `{${props.name}}${children.join('')}{/${props.name}}`;
        } else {
            return `{${props.name}}`;
        }
    }
    return children.map((child) => {
        return typeof child === 'string' ? child : jsonToText(...child);
    }).join('');
};


const renderStringTranslation = (translation) => {
    if (typeof translation === 'string') {
        return translation;
    } else {
        // we do not support placeholders in Translate.string(), but
        // when compiling a translation dict to JSON we cannot know in
        // which context a string is used, so we have to replace the params
        // with the original placeholders
        return jsonToText(...translation);
    }
};


export class Singular extends React.Component {
    static propTypes = {
        children: PropTypes.any.isRequired,
    };

    render() {
        const {children} = this.props;
        return children;
    }
}


export class Plural extends React.Component {
    static propTypes = {
        children: PropTypes.any.isRequired,
    };

    render() {
        const {children} = this.props;
        return children;
    }
}


const getGettextFuncs = (args) => {
    let gettext, ngettext, pgettext, npgettext;
    if (args.length === 1) {
        const [obj] = args;
        gettext = obj.gettext.bind(obj);
        ngettext = obj.ngettext.bind(obj);
        pgettext = obj.pgettext ? obj.pgettext.bind(obj) : undefined;
        npgettext = obj.npgettext ? obj.npgettext.bind(obj) : undefined;
    } else if (args.length === 2) {
        [gettext, ngettext] = args;
    } else if (args.length === 4) {
        [gettext, ngettext, pgettext, npgettext] = args;
    } else {
        throw new Error('Expected object containing gettext/ngettext(/pgettext/npgattext) or 2/4 args with the funcs');
    }

    if (!pgettext) {
        pgettext = (ctx, ...params) => gettext(...params);
    }
    if (!npgettext) {
        npgettext = (ctx, ...params) => ngettext(...params);
    }

    return {gettext, ngettext, pgettext, npgettext};
};


const pickGettextFunc = (context, gettext, pgettext) => {
    if (context) {
        return (...args) => pgettext(context, ...args);
    } else {
        return gettext;
    }
};


const getParamValues = (element) => {
    const obj = Object.create(null);
    React.Children.forEach(element.props.children, (child) => {
        if (React.isValidElement(child) && child.type === Param) {
            obj[child.props.name] = child.props.value;
        }
    });
    return obj;
};


export const makeComponents = (...args) => {
    const {gettext, ngettext, pgettext, npgettext} = getGettextFuncs(args);

    class Translate extends React.Component {
        static propTypes = {
            children: PropTypes.any.isRequired,
            context: PropTypes.string,
        };

        static defaultProps = {
            context: undefined,
        };

        static string(string, context = undefined) {
            const gettextFunc = pickGettextFunc(context, gettext, pgettext);
            return renderStringTranslation(gettextFunc(string));
        }

        constructor(props) {
            super(props);
            this.original = getTranslatableString(props.children);
        }

        render() {
            const {children, context} = this.props;
            const gettextFunc = pickGettextFunc(context, gettext, pgettext);
            const translation = gettextFunc(this.original);
            if (translation === this.original) {
                // if there's no translation gettext gives us the input string
                // which does not contain the information needed to render it!
                return children;
            }
            return renderTranslation(translation, getParamValues(this));
        }
    }


    class PluralTranslate extends React.Component {
        static propTypes = {
            children: PropTypes.any.isRequired,
            count: PropTypes.number.isRequired,
            context: PropTypes.string,
        };

        static defaultProps = {
            context: undefined,
        };

        static string(singular, plural, count = 1, context = undefined) {
            const gettextFunc = pickGettextFunc(context, ngettext, npgettext);
            return renderStringTranslation(gettextFunc(singular, plural, count));
        }

        constructor(props) {
            super(props);
            React.Children.forEach(props.children, (child) => {
                if (!React.isValidElement(child)) {
                    throw new Error(`Unexpected PluralTranslate child: ${child}`);
                } else if (child.type === Singular) {
                    this.singularString = getTranslatableString(child.props.children);
                } else if (child.type === Plural) {
                    this.pluralString = getTranslatableString(child.props.children);
                }
            });
        }

        getChild(plural) {
            const {children} = this.props;
            const component = plural ? Plural : Singular;
            for (const child of React.Children.toArray(children)) {
                if (React.isValidElement(child) && child.type === component) {
                    return child;
                }
            }
        }

        render() {
            const {count, context} = this.props;
            const gettextFunc = pickGettextFunc(context, ngettext, npgettext);
            const translation = gettextFunc(this.singularString, this.pluralString, count);
            if (translation === this.singularString) {
                return this.getChild(false);
            } else if (translation === this.pluralString) {
                return this.getChild(true);
            }
            const values = getParamValues(this.getChild(count !== 1));
            return renderTranslation(translation, values);
        }
    }


    return {Translate, PluralTranslate};
};
