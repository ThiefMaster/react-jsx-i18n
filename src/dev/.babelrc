{
  "plugins": [
    "@babel/plugin-proposal-object-rest-spread"
  ],
  "presets": [
    [
      "@babel/env",
      {
        "targets": {
          "node": "current"
        }
      }
    ]
  ]
}
