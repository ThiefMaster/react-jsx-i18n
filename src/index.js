export {makeComponents, Singular, Plural, Param} from './components';
export {default as makeI18nPlugin} from './extract-plugin';
export {default as poToReact} from './po2react';
