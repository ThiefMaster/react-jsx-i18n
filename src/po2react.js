import * as fs from 'fs';
import gettextParser from 'gettext-parser';


const jsonifyMessage = (message) => {
    if (!/{[a-z]+}/.test(message)) {
        return message;
    }

    const json = ['Fragment', null];
    const parts = message.split(/({\/?[a-z]+})/).filter((part) => !!part);
    for (let i = 0; i < parts.length; i++) {
        const match = parts[i].match(/^{([a-z]+)}$/);
        if (!match) {
            // plain string
            json.push(parts[i]);
            continue;
        }
        const paramName = match[1];
        if (parts[i + 2] === `{/${paramName}}`) {
            // param with a body
            json.push(['Param', {name: paramName}, parts[i + 1]]);
            i += 2;
        } else {
            // param with no body
            json.push(['Param', {name: paramName}]);
        }
    }
    return json;
};


const poToReact = (poFile, domain = 'messages') => {
    const po = gettextParser.po.parse(fs.readFileSync(poFile), 'UTF-8');

    const translations = Object.values(po.translations['']).map((item) => {
        return {[item.msgid]: item.msgstr.map(jsonifyMessage)};
    });

    return {
        [domain]: {
            ...Object.assign(Object.create(null), ...translations),
            '': {
                domain,
                lang: po.headers.language,
                plural_forms: po.headers.plural_forms,
            }
        }
    };

};

export default poToReact;
