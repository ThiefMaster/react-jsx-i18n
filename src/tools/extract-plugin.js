import cleanJSXElementLiteralChild from '@babel/types/lib/utils/react/cleanJSXElementLiteralChild';


const collapseWhitespace = (string) => {
    // for translated strings we never want consecutive whitespace
    return string.replace(/\s+/g, ' ');
};


const processText = (path) => {
    const result = [];
    cleanJSXElementLiteralChild(path.node, result);
    return result.length ? result[0].value : '';
};


const processParam = (path) => {
    const childElement = path.node.openingElement;
    const paramName = childElement.attributes.filter((attr) => (
        attr.name.name === 'name'
    ))[0].value.value;
    const children = path.get('children');
    if (children.length === 0) {
        return `{${paramName}}`;
    } else if (children.length !== 1) {
        throw children[1].buildCodeFrameError(`Param has too many children (${children.length}), expected max. 1`);
    }
    const childPath = children[0];
    if (childPath.type !== 'JSXText') {
        throw childPath.buildCodeFrameError(`Unexpected Param child node: ${childPath.type}`);
    }
    const body = processText(childPath);
    return `{${paramName}}${body}{/${paramName}}`;
};


const processExpression = (path, expression, types) => {
    if (types.isStringLiteral(expression)) {
        return expression.value;
    } else if (types.isBinaryExpression(expression)) {
        if (expression.operator !== '+') {
            throw path.buildCodeFrameError(
                `Expected '+' operator in binary expression; found '${expression.operator}' instead`
            );
        }
        return processExpression(path, expression.left, types) + processExpression(path, expression.right, types);
    }
    throw path.buildCodeFrameError(
        `Expressions may only contain a string literal or a concatenation of them; found ${expression.type} instead`
    );
};


const processTranslatableElement = (path, types) => {
    const elementName = path.node.openingElement.name.name;
    const stringParts = path.get('children').map((childPath) => {
        if (childPath.type === 'JSXText') {
            return processText(childPath);
        } else if (childPath.type === 'JSXExpressionContainer') {
            return processExpression(childPath, childPath.node.expression, types);
        } else if (childPath.type === 'JSXElement') {
            const childElement = childPath.node.openingElement;
            if (childElement.name.name !== 'Param') {
                throw childPath.buildCodeFrameError(`Unexpected ${elementName} child tag: ${childElement.name.name}`);
            }
            return processParam(childPath);
        } else {
            throw childPath.buildCodeFrameError(`Unexpected ${elementName} child node: ${childPath.type}`);
        }
    });
    return collapseWhitespace(stringParts.join('')).trim();
};


const getLocation = (path, state) => {
    return `${state.file.opts.filename}:${path.node.loc.start.line}`;
};


const getContext = (path) => {
    const element = path.node.openingElement;
    const contextAttr = element.attributes.filter((attr) => attr.name.name === 'context')[0];
    return contextAttr ? contextAttr.value.value : undefined;
};


const processTranslate = (path, state, types) => {
    const translatableString = processTranslatableElement(path, types);
    return {
        msgid: translatableString,
        msgctxt: getContext(path),
        reference: getLocation(path, state),
    };
};


const processTranslateString = (path, state, funcName, types) => {
    const args = path.node.arguments;
    if (args.length === 0) {
        throw path.buildCodeFrameError('Translate.string() called with no arguments');
    }
    const msgid = processExpression(path, args[0], types);
    const msgctxt = args[1] ? processExpression(path, args[1], types) : undefined;
    return {
        msgid,
        msgctxt,
        reference: getLocation(path, state),
    };
};


const processPluralTranslate = (path, state, types) => {
    let singularPath, pluralPath;
    path.get('children').filter((x) => x.node.type === 'JSXElement').forEach((childPath) => {
        const element = childPath.node.openingElement;
        const elementName = element.name.name;
        if (elementName === 'Singular') {
            if (singularPath) {
                throw childPath.buildCodeFrameError('More than one Singular tag found');
            }
            singularPath = childPath;
        } else if (elementName === 'Plural') {
            if (pluralPath) {
                throw childPath.buildCodeFrameError('More than one Plural tag found');
            }
            pluralPath = childPath;
        } else {
            throw childPath.buildCodeFrameError(`Unexpected PluralTranslate child tag: ${elementName}`);
        }
    });
    if (!singularPath) {
        throw path.buildCodeFrameError('No Singular tag found');
    }
    if (!pluralPath) {
        throw path.buildCodeFrameError('No Plural tag found');
    }
    return {
        msgid: processTranslatableElement(singularPath, types),
        msgid_plural: processTranslatableElement(pluralPath, types),
        msgctxt: getContext(path),
        reference: getLocation(path, state),
    };
};


const processPluralTranslateString = (path, state, funcName, types) => {
    const args = path.node.arguments;
    if (args.length < 2) {
        throw path.buildCodeFrameError('PluralTranslate.string() called with less than 2 arguments');
    }
    const msgid = processExpression(path, args[0], types);
    const msgid_plural = processExpression(path, args[1], types);  // eslint-disable-line camelcase
    const msgctxt = args[3] ? processExpression(path, args[3], types) : undefined;
    return {
        msgid,
        msgid_plural,
        msgctxt,
        reference: getLocation(path, state),
    };
};


const makeI18nPlugin = () => {
    const entries = [];
    const i18nPlugin = ({types}) => {
        return {
            visitor: {
                JSXElement(path, state) {
                    const elementName = path.node.openingElement.name.name;
                    if (elementName === 'Translate') {
                        entries.push(processTranslate(path, state, types));
                    } else if (elementName === 'PluralTranslate') {
                        entries.push(processPluralTranslate(path, state, types));
                    }
                },
                CallExpression(path, state) {
                    const callee = path.node.callee;
                    if (callee.type !== 'MemberExpression' || callee.object.type !== 'Identifier') {
                        return;
                    }
                    const elementName = callee.object.name;
                    if (elementName !== 'Translate' && elementName !== 'PluralTranslate') {
                        return;
                    }
                    const property = callee.property;
                    const funcName = callee.computed ? property.value : property.name;
                    if (callee.computed && property.type !== 'StringLiteral') {
                        return;
                    } else if (!callee.computed && property.type !== 'Identifier') {
                        return;
                    } else if (funcName !== 'string') {
                        return;
                    }
                    // we got a proper call of one of our translation functions
                    const qualifiedFuncName = `${elementName}.${funcName}`;
                    if (elementName === 'Translate') {
                        entries.push(processTranslateString(path, state, qualifiedFuncName, types));
                    } else if (elementName === 'PluralTranslate') {
                        entries.push(processPluralTranslateString(path, state, qualifiedFuncName, types));
                    }
                }
            }
        };
    };

    return {i18nPlugin, entries};
};

export default makeI18nPlugin;
