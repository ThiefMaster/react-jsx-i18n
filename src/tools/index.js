export {default as makeI18nPlugin} from './extract-plugin';
export {default as poToReact} from './po2react';
