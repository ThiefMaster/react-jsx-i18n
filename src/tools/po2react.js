import fs from 'fs';
import gettextParser from 'gettext-parser';


export const jsonifyMessage = (message) => {
    if (!/{[a-zA-Z_]+}/.test(message)) {
        return message;
    }

    const json = ['Fragment', null];
    const parts = message.split(/({\/?[a-zA-Z_]+})/).filter((part) => !!part);
    for (let i = 0; i < parts.length; i++) {
        const match = parts[i].match(/^{([a-zA-Z_]+)}$/);
        if (!match) {
            // plain string
            json.push(parts[i]);
            continue;
        }
        const paramName = match[1];
        if (parts[i + 2] === `{/${paramName}}`) {
            // param with a body
            json.push(['Param', {name: paramName}, parts[i + 1]]);
            i += 2;
        } else {
            // param with no body
            json.push(['Param', {name: paramName}]);
        }
    }
    return json;
};


const poToReact = (poFile, domain = 'messages') => {
    const po = gettextParser.po.parse(fs.readFileSync(poFile), 'UTF-8');

    const translations = Object.create(null);
    Object.values(po.translations).forEach((items) => {
        Object.values(items).forEach((item) => {
            const msgid = item.msgctxt ? `${item.msgctxt}\u0004${item.msgid}` : item.msgid;
            translations[msgid] = item.msgstr.map(jsonifyMessage);
        });
    });

    return {
        [domain]: {
            ...translations,
            '': {
                domain,
                lang: po.headers.language,
                plural_forms: po.headers['plural-forms'],
            }
        }
    };
};

export default poToReact;
