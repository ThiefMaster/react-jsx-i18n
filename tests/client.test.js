/* global test:false, expect:false */

import React from 'react';
import renderer from 'react-test-renderer';

import {makeComponents, Param, Singular, Plural} from '../src/client';
import {jsonifyMessage} from '../src/tools/po2react';


const expectJSX = (jsx) => expect(renderer.create(jsx).toJSON());
const expectJSXWrapper = (jsx) => expect(() => renderer.create(jsx).toJSON());
const gettext = expression => jsonifyMessage(`translated-${expression}`);
const ngettext = (sExpression, pExpression, n) => jsonifyMessage(n === 0 || n > 1 ? `plural-${pExpression}` : `singular-${sExpression}`);

const {Translate, PluralTranslate} = makeComponents(gettext, ngettext);

// This prevents printing warnings to the console about missing error boundaries
beforeEach(() => {
    jest.spyOn(console, 'error');
    global.console.error.mockImplementation(() => {});
});

afterEach(() => {
    global.console.error.mockRestore();
});


test('Basic translation works', () => {
    expectJSX(<Translate>kajak</Translate>).toBe('translated-kajak');
    expectJSX(<Translate>Fetchez la vache</Translate>).toBe('translated-Fetchez la vache');
    expectJSX(<Translate>Fetchez la vache <Param name="number" value={5} /></Translate>).toEqual(expect.arrayContaining(['translated-Fetchez la vache ', '5']));
    expectJSX(Translate.string('kajak')).toBe('translated-kajak');
});

test('Pluralized translation works', () => {
    const message = count => (
        <PluralTranslate count={count}>
            <Singular>
                This is a vache
            </Singular>
            <Plural>
                This is <Param name="count" value={count} /> vaches
            </Plural>
        </PluralTranslate>
    );

    expectJSX(message(1)).toBe('singular-This is a vache');
    expectJSX(message(2)).toEqual(expect.arrayContaining(['plural-This is ', '2', ' vaches']));

    const secondMessage = count => (
        <PluralTranslate count={count}>
            <Singular>
                fetchez la vache
            </Singular>
            <Plural>
                FETCHEZ LES VACHES
            </Plural>
        </PluralTranslate>
    );

    expectJSX(secondMessage(1)).toBe('singular-fetchez la vache');
    expectJSX(secondMessage(2)).toBe('plural-FETCHEZ LES VACHES');
    expectJSX(PluralTranslate.string('Fetchez la Vache', 'Fetchez les vaches', 1)).toBe('singular-Fetchez la Vache');
    expectJSX(PluralTranslate.string('Fetchez la Vache', 'Fetchez les vaches', 2)).toBe('plural-Fetchez les vaches');
});

test('JSX expressions work', () => {
    expectJSX(<Translate>This is a {'test'}</Translate>).toBe('translated-This is a test');
    expectJSXWrapper(<Translate>5 + 5 = {5 + 5}</Translate>).toThrow();
});

test('Translate component can only contain elements of specific type', () => {
    expectJSXWrapper(<Translate>Test <Translate>test</Translate></Translate>).toThrow();
    expectJSXWrapper(<Translate>My name is <Param name="name" value="Michał" /> <Param name="name" value="Michał" /></Translate>).toThrow();
});

test('PluralTranslate component can only contain elements of specific type', () => {
    const element = (
        <PluralTranslate count={5}>
            {'That is a test'}
        </PluralTranslate>
    );

    expectJSXWrapper(element).toThrow();
});

test('Whitespaces are handled correctly', () => {
    const element = (
        <Translate>
            Fetchez   la    vache


            Fetchez la vache
        </Translate>
    );

    expectJSX(element).toBe('translated-Fetchez la vache Fetchez la vache');

    const secondElement = (
        <Translate>
            Fetchez la     <Param name="vache" value="vache" />
            <Param name="random" value="test" />{' '}
            <Param name="count" value={5} />
        </Translate>
    );

    expectJSX(secondElement).toEqual(expect.arrayContaining(['translated-Fetchez la ', 'vache', 'test', ' ', '5']));

    const pluralMessage = count => (
        <PluralTranslate count={count}>
            <Singular>
                fetchez la vache <Param name="number" value={count} />
            </Singular>
            <Plural>
                FETCHEZ LES VACHES <Param name="number" value={count} />
            </Plural>
        </PluralTranslate>
    );

    expectJSX(pluralMessage(1)).toEqual(expect.arrayContaining(['singular-fetchez la vache ', '1']));
    expectJSX(pluralMessage(2)).toEqual(expect.arrayContaining(['plural-FETCHEZ LES VACHES ', '2']));
});
