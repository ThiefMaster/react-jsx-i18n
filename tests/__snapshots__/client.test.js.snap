// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example component works properly with translations 1`] = `
<div>
  Hallo & Welt
  <br />
  Hey 
  dude
  , du willst 
  <a
    onClick={[Function]}
  >
    mich anklicken
  </a>
   und weißt das auch!
  <br />
  Tschüss Welt
  <br />
  Tschüss Welt
  <br />
  Etwas &lt;HTML&gt; & Entities: → &
  <br />
  <span
    title="\\"<strong>Ratten.</strong>\\""
  >
    Hover mich
  </span>
  <br />
  &lt;HTML&gt; wird unescaped extracted
  <br />
  xxx foo bar
  1
  2
   muh
  <hr />
  Du hast 
  1
   Ratte.
  <br />
  Du hast 
  0
   Ratten.
  <br />
  Du hast 
  1
   Ratte.
  <br />
  Du hast 
  2
   Ratten.
  <br />
  Junge Katzen:
   
  Kätzchen
  <br />
  Junge Hunde:
   
  Welpe
  <br />
  Das ist ein hervorgehobener dynamischer Wert: 
  <em>
    1000
  </em>
  Das ist ein hervorgehobener übersetzter Wert: 
  <em>
    hallo
  </em>
</div>
`;

exports[`Example component works properly without translations 1`] = `
<div>
  Hello & World
  <br />
  Hey 
  dude
  , you want to 
  <a
    onClick={[Function]}
  >
    click me
  </a>
   and you know it!
  <br />
  Bye World
  <br />
  Bye World
  <br />
  Some &lt;HTML&gt; & entities: → &
  <br />
  <span
    title="\\"<strong>Rats.</strong>\\""
  >
    Hover me
  </span>
  <br />
  &lt;HTML&gt; is unescaped when extracted
  <br />
  xxx foo bar
  1
  2
   
  moo
  <hr />
  You have 
  1
   rat.
  <br />
  You have 
  0
   rats.
  <br />
  You have 
  1
   rat.
  <br />
  You have 
  2
   rats.
  <br />
  Little cats:
   
  offspring
  <br />
  Little dogs:
   
  offspring
  <br />
  This is an emphasized dynamic value: 
  <em>
    1000
  </em>
  This is an emphasized translated value: 
  <em>
    hello
  </em>
</div>
`;

exports[`Example strings work properly with translations 1`] = `
Object {
  "bigCat1": "KATZE",
  "bigCat2": "KATZEN",
  "cat1": "Katze",
  "cat2": "Katzen",
  "meow": "Kätzchen",
  "oneParam": "Foo: bar",
  "twoParams": "Foo: bar",
  "ugly": "Du    bist hässlich!",
  "v1": "\\"<strong>Ratten.</strong>\\"",
  "v2": "\\"<strong>Ratten.</strong>\\"",
  "v3": "FooBar",
  "v4": "FooBar",
  "weird": "Foo   stuff Bar world123 XXX",
}
`;

exports[`Example strings work properly without translations 1`] = `
Object {
  "bigCat1": "cat",
  "bigCat2": "cats",
  "cat1": "cat",
  "cat2": "cats",
  "meow": "offspring",
  "oneParam": "foo: bar",
  "twoParams": "foo: bar",
  "ugly": "You    are ugly!",
  "v1": "\\"<strong>Rats.</strong>\\"",
  "v2": "\\"<strong>Rats.</strong>\\"",
  "v3": "foobar",
  "v4": "foobar",
  "weird": "foo   stuff bar world123 xxx",
}
`;

exports[`Prop updates are working with translation 1`] = `
<div>
  translated-Hello, 
  user
  <br />
  plural-You have 
  0
   new messages.
</div>
`;

exports[`Prop updates are working with translation 2`] = `
<div>
  translated-Hello, 
  world
  <br />
  plural-You have 
  0
   new messages.
</div>
`;

exports[`Prop updates are working with translation 3`] = `
<div>
  translated-Hello, 
  important dude
  <br />
  singular-You have a new message.
</div>
`;

exports[`Prop updates are working without translation 1`] = `
<div>
  Hello, 
  user
  <br />
  You have 
  0
   new messages.
</div>
`;

exports[`Prop updates are working without translation 2`] = `
<div>
  Hello, 
  world
  <br />
  You have 
  0
   new messages.
</div>
`;

exports[`Prop updates are working without translation 3`] = `
<div>
  Hello, 
  important dude
  <br />
  You have a new message.
</div>
`;
